### ISSUE TITLE
False positive with operator+ and lists (no-member)

### ISSUE THREAD

**Issue by hemberger** [PyCQA/pylint#4826]
### Bug description

When using `operator+` to merge two lists (in the context of a class method?), pylint incorrectly thinks that the objects in the combined list are also of type `list`.

The simplest example I can construct is the following:

```python
# pylint: disable=C0103,C0114,C0115,C0116,R0903

class A:
    def __init__(self):
        self.Alist = []
        self.data = []

    def run(self):
        for a in [self] + self.Alist:
            for _ in a.data:  # # <-- causes "no-member" warning
                pass
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:12:21: E1101: Instance of 'list' has no 'data' member (no-member)
```

### Expected behavior

I would expect there to be no warning, since both `self` and `self.Alist` are (nominally) of type `A`, which does have a member called `data`.

If instead of using `[self] + self.Alist`, any of the following will _not_ result in a warning:
* `for a in [self]`
* `for a in self.Alist`

### Pylint version

```shell
pylint 2.9.6
astroid 2.6.5
Python 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
```

### OS / Environment

CentOS 8.4

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening this issue. I'm classifying as minor since

```python
class A:
    def __init__(self):
        self.Alist = [A()]
        self.data = []

    def run(self):
        for a in [self] + self.Alist:
            for _ in a.data:  # # <-- causes "no-member" warning
                pass

A().run()
```
Gives no pylint warning and putting a string instead of ``A()`` gives a proper pylint warning.

**Comment by DanielNoord** [comment]
This is probably an issue in `astroid`.
I wrote the following code to show different behaviour in similar code:
```python
import astroid

module = astroid.parse(
    """
for a in [1] + [2]:
    print(a)
"""
)

print(list(module.body[0].body[0].value.args[0].infer()))

module = astroid.parse(
    """
class A:
    def __init__(self):
        for a in [self] + []:
            print(a)
"""
)

print(list(module.body[0].body[0].body[0].body[0].value.args[0].infer()))

module = astroid.parse(
    """
class A:
    def __init__(self):
        for a in [self, 1]:
            print(a)
"""
)

print(list(module.body[0].body[0].body[0].body[0].value.args[0].infer()))
```
```console
❯ python3 test.py
[<Const.int l.2 at 0x1075c9060>, <Const.int l.2 at 0x1075c90f0>]
[<Instance of builtins.list at 0x4418479552>]
[<Instance of .A at 0x4418481136>, <Const.int l.4 at 0x1075c8400>]
```

With `for a in [self] + []` the inference seems to stop at recognising that it is a list and is unable to go any further.
I haven't looked at any `astroid` internals yet, but this should be a good start to find where this needs to be fixed or handled better.
